src/auth/downscopedclient.ts

```
@@ -0,0 +1,254 @@
+    if (cab.accessBoundary.accessBoundaryRules.length === 0) {
+      throw new Error('At least one access boundary rule needs to be defined.');
+    } else if (cab.accessBoundary.accessBoundaryRules.length > 10) {
+      throw new Error('Access boundary rule exceeds limit, max 10 allowed.');
+    }
+
+    // Check at least one permission should be defined in each access boundary rule.
+    for (const rule of cab.accessBoundary.accessBoundaryRules) {
+      if (rule.availablePermissions.length === 0) {
+        throw new Error(
+          'At least one permission should be defined in access boundary rules.'
+        );
+      }
+    }
+
+    this.stsCredential = new sts.StsCredentials(STS_ACCESS_TOKEN_URL);
+    // Default OAuth scope. This could be overridden via public property.
+    this.cachedDownscopedAccessToken = null;
+    this.credentialAccessBoundary = cab;
+    this.authClient = client;
+    // As threshold could be zero,
+    // eagerRefreshThresholdMillis || EXPIRATION_TIME_OFFSET will override the
+    // zero value.
+    if (typeof additionalOptions?.eagerRefreshThresholdMillis !== 'number') {
+      this.eagerRefreshThresholdMillis = EXPIRATION_TIME_OFFSET;
+    } else {
+      this.eagerRefreshThresholdMillis = additionalOptions!
+        .eagerRefreshThresholdMillis as number;
+    }
+    this.forceRefreshOnFailure = !!additionalOptions?.forceRefreshOnFailure;
+  }
+
+  /**
+   * Provides a mechanism to inject Downscoped access tokens directly.
+   * When the provided credential expires, a new credential, using the
+   * external account options are retrieved.
+   * Notice DownscopedClient is the broker class mainly used for generate
+   * downscoped access tokens, it is unlikely we call this function in real
+   * use case.
+   * We implement to make this a helper function for testing all cases in getAccessToken().
+   * @param credentials The Credentials object to set on the current client.
+   */
+  setCredentials(credentials: Credentials) {
```

To make things simple. Let's require the provided credentials to have an expiration time. Can you throw an error if they don't?